asutherland/ni-controllers-lib
# Add WebUSB support

#### What does this PR do?
It refactors the USB access code to enable the user to pick the runtime implementation they need.
So now this library can be used equally in Node or the Web platform. Disclaimer: Only the _Blink_ like browsers are supported because it requires WebUSB and WebHID (as of today still behind a flag: cf. chrome://flags/#enable-experimental-web-platform-features).

Because the Web version of the USB api's are asynchronous by design I had to introduce an `init` phase. So for example, to start the node version we should now do:

```ts
// web would be new MaschineMk3(createWebHidAdapter, createWebUsbAdapter)
const mk3 = new MaschineMk3(createNodeHidAdapter, createNodeUsbAdapter)
await mk3.init()
```

#### How should this be tested by the reviewer?
I reused the test app code for the mk3 and created an examples folder with bot boostraps snippets (Node and Web). They can be respectively run with:

```bash
npm run example:maschinemk3:node
```
and (the web version runs parcel bundler and you can then go to http://localhost:1234)
```bash
npm run example:maschinemk3:web
```
They should deliver the same results

#### What are the relevant tickets?
https://github.com/asutherland/node-traktor-f1/issues/2
